### PR TITLE
Add scripts to automate GitHub releases

### LINKED ISSUE #3403: Create automated Github Releases deploy script

### PR DESCRIPTION
**Pull request by whymarrh**
This PR updates our CircleCI configuration, adding new 1 new job and 1 new workflow to automate parts of our release process. The work here ultimately results in a GitHub Release with the artifacts from `npm run dist` attached.

### The Workflow

The high-level workflow is:

- A version branch is created off of `develop`
    1. CI sets the manifest version and commits that
    2. CI creates a Pull Request for the version
- When the Pull Request gets merged, CI creates a tag and GitHub Release for the commit

I've been playing around with the configuration on my fork, if you want to look at the [CI builds](https://circleci.com/gh/whymarrh/workflows/metamask-extension) for that repository:

- An example set of version branch builds: https://circleci.com/workflow-run/375c708c-9ae9-4264-9815-0ead4e3055b6
- An example `develop` build: https://circleci.com/workflow-run/83240798-7fea-4d4c-811f-e45cb2202618

#### Step 1: Creating a RC Pull Request

**End result:** a pull request for the version is created

**Process:** when a branch is created matching `^Version-v\d+[.]\d+[.]\d+`, a CI job will use the version number from the branch name to update the version field in the manifest, and create a PR for that change. That PR can then be reviewed and updated as appropriate. When we go to merge the PR, we can squash+merge the changes into a single commit with the message `Version vX.Y.Z`.

#### Step 2: Creating a GitHub Release

**End result:** a release is created with the Chrome and Firefox builds attached

**Process:** when a commit to `develop` is made with a message matching `^Version v\d+[.]\d+[.]\d+`. the CI build for that commit will include tagging it as `vX.Y.Z` and creating a release with the Chrome and Firefox builds as assets.

### ~Unanswered~ Questions

- [x] ~How do we want to deal with `CHANGELOG.md`?~ Scripts added to update the changelog
- [x] ~Who should we set as reviewers on the RC PR?~ `@MetaMask/extension-release-team`

## Tasks

- [x] Set `GITHUB_TOKEN` and `GITHUB_TOKEN_USER` environment variables in CircleCI
- [x] Double-check copy

### Refs

- Closes https://github.com/MetaMask/metamask-extension/issues/3403
- [Using Workflows to Schedule Jobs](https://circleci.com/docs/2.0/workflows/)
- [Configuring CircleCI](https://circleci.com/docs/2.0/configuration-reference/)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -17,19 +29,12 @@ workflows:
...
-      # - test-deps:
-      #     requires:
-      #       - prep-deps-npm
       - test-e2e-chrome:
           requires:
             - prep-deps-npm
       - test-e2e-firefox:
           requires:
             - prep-deps-npm
-      # - test-e2e-beta-drizzle:
```
Comment (danjm) [inline .circleci/config.yml]: why are we removing the drizzle tests as part of this PR?

File: .circleci/scripts/release-create-gh-release [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+current_commit_msg=$(git show -s --format='%s' HEAD)
+
+if grep --quiet '^Version v' <<< "$current_commit_msg"
+then
+    printf '%s\n' 'Creating GitHub Release'
+    tag="${current_commit_msg/Version /}"
+    pushd builds
+        hub release create \
+            --attach metamask-chrome-6.5.3.zip \
```
Comment (danjm) [inline .circleci/scripts/release-create-gh-release]: I guess the `6.5.3` needs to be set dynamically?

**Review by danjm — DISMISSED** [review]

**Comment by danjm** [comment]
It would be nice to have an npm command that automates "Step 1: Creating a RC Pull Request". Should that fall within the scope of this PR?

File: .circleci/config.yml (hunk 1 above)
Comment (whymarrh) [reply .circleci/config.yml]: That's a good catch, I'll add these back

File: .circleci/scripts/release-create-gh-release (hunk 2 above)
Comment (whymarrh) [reply .circleci/scripts/release-create-gh-release]: That's a good point, we can peek into the directory an pick out the files

**Comment by whymarrh** [comment]
> It would be nice to have an npm command that automates "Step 1: Creating a RC Pull Request". Should that fall within the scope of this PR?

I don't think there's much there. We can literally create the branch on GitHub itself in a second.

File: .circleci/scripts/release-create-gh-release [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    curl -sSL 'https://github.com/github/hub/releases/download/v2.11.2/hub-linux-amd64-2.11.2.tgz' | tar xz
+    PATH="$PATH:$PWD/hub-linux-amd64-2.11.2/bin"
+popd
+
+current_commit_msg=$(git show -s --format='%s' HEAD)
+
+if grep --quiet '^Version v' <<< "$current_commit_msg"
+then
+    printf '%s\n' 'Creating GitHub Release'
+    tag="${current_commit_msg/Version /}"
```
Comment (whymarrh) [inline .circleci/scripts/release-create-gh-release]: We should use the 2nd word here as the version, to allow us to always squash+merge, e.g.

```
Version v1.2.3 RC (#foo)
```

File: .circleci/config.yml [hunk 4]
```diff
@@ -2,12 +2,24 @@ version: 2
...
 workflows:
   version: 2
-  full_test:
+  test_and_release:
     jobs:
+      - create_release_pull_request:
+          filters:
+            branches:
+              only:
+                - /^Version-v(\d+)[.](\d+)[.](\d+)/
```
Comment (danfinlay) [inline .circleci/config.yml]: This only runs upon merge of a version branch?

File: .circleci/scripts/release-create-gh-release [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    exit 1
+fi
+
+printf '%s\n' 'Installing hub CLI'
+pushd "$(mktemp -d)"
+    curl -sSL 'https://github.com/github/hub/releases/download/v2.11.2/hub-linux-amd64-2.11.2.tgz' | tar xz
+    PATH="$PATH:$PWD/hub-linux-amd64-2.11.2/bin"
+popd
+
+current_commit_msg=$(git show -s --format='%s' HEAD)
```
Comment (danfinlay) [inline .circleci/scripts/release-create-gh-release]: Does this amount to the changelog?

Our current changelog has a weird bug where it creates a number of "changelog-description" entries, and I usually end up manually removing those. We might want to also address that too, to achieve full (clean) automation, since having a good changelog is really nice.

**Review by danfinlay — DISMISSED** [review]
A couple small questions.

Overall, looks awesome, definitely a few flags & commands that are new to me here.

Looks like it does stop short of automating the publication to the stores, which makes sense for now, as that's the most security sensitive bit. Since this PR excludes _the most security critical piece_, it seems generally safe to try out and see how we like it!

File: .circleci/config.yml (hunk 4 above)
Comment (whymarrh) [reply .circleci/config.yml]: This instructs CircleCI to only run the `create_release_pull_request` job on branches matching this pattern. That is, any commits to a version branch will trigger this job.

File: .circleci/scripts/release-bump-manifest-version [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+if [[ "${CIRCLECI:-}" != 'true' ]]
+then
+    printf '%s\n' 'CIRCLECI environment variable must be set to true'
+    exit 1
+fi
+
+printf '%s\n' 'Updating the manifest version if needed'
+
+version="${CIRCLE_BRANCH/Version-v/}"
+cat <<< "$(jq ".version = \"$version\"" app/manifest.json)" > app/manifest.json
```
Comment (Gudahtt) [inline .circleci/scripts/release-bump-manifest-version]: Shellcheck shows a warning for this line (https://github.com/koalaman/shellcheck/wiki/SC2094)

Generally speaking, reading and writing to the same file in a pipeline seems like a bad idea. But I'm having difficulty imaging what could go wrong in this specific case.

File: .circleci/scripts/release-bump-manifest-version [hunk 7]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
+
+set -e
+set -u
+set -o pipefail
+
+if [[ "${CI:-}" != 'true' ]]
```
Comment (Gudahtt) [inline .circleci/scripts/release-bump-manifest-version]: I noticed the `${VAR:-}` parameter expansion being used throughout these scripts, and I'm not entirely sure what it does.

I am familiar with `${VAR:-foo}`, where `foo` would be substituted if `$VAR` was unset or null. But `${VAR:-}` would just substitute an empty string I suppose? Which seems irrelevant in all of the cases I've seen so far.

File: .circleci/scripts/release-create-gh-release [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    exit 1
+fi
+
+if [[ "${CIRCLECI:-}" != 'true' ]]
+then
+    printf '%s\n' 'CIRCLECI environment variable must be set to true'
+    exit 1
+fi
+
+printf '%s\n' 'Installing hub CLI'
```
Comment (Gudahtt) [inline .circleci/scripts/release-create-gh-release]: This step should probably be inside the `if` block below, for when a release commit is detected. Otherwise this would run on each commit to the `develop` branch, wouldn't it?

**Review by Gudahtt — CHANGES_REQUESTED** [review]

File: .circleci/scripts/release-create-gh-release (hunk 5 above)
Comment (Gudahtt) [reply .circleci/scripts/release-create-gh-release]: This is looking at specifically the version commit made on line 30 of the `release-bump-manifest-version` script above, not any of the commits in between this version and the last.

I took a look at the changelog script, and I think I know what's causing the problem (it's splitting on newlines in the commit body, treating each line as a separate commit). I can put up a separate PR to fix that, and maybe add some error handling.

Once that's fixed, perhaps the changelog can be updated as part of the `release-bump-manifest-version` script.

File: .circleci/scripts/release-bump-manifest-version (hunk 7 above)
Comment (Gudahtt) [reply .circleci/scripts/release-bump-manifest-version]: Whoops, nevermind, I get it. It's because of `set -u`. I haven't used that in a while.

File: .circleci/scripts/release-create-gh-release (hunk 5 above)
Comment (Gudahtt) [reply .circleci/scripts/release-create-gh-release]: I've made these changes here: https://github.com/MetaMask/metamask-extension/pull/6753

It's still not at the point where I'd trust it enough to include it in `release-bump-manifest-version` unfortunately. The commit messages are too inconsistent. We could certainly improve that, but that would likely introduce additional friction for contributors, which I'd rather avoid.

I might have an idea for how to accomplish this without adding much friction. We could use the squash-and-merge strategy for all PRs (e.g. disable the other two options). That'd ensure that the commit messages at least have _roughly_ the same format, at least as long as people stick with the defaults populated by GitHub (which adds `(#<PR>)` to the end of the commit subject).

We could take it a step further by adding a GitHub App to enforce certain things about the commit message, and make helpful suggestions. e.g. a commit linter, but only for the final squashed commit, not for the individual feature branch commits. I've seen something like that before - there might be an existing app that does this already, or at least approximates it. That also makes it easier to adopt additional conventions, like categorizing each commit. That would let us generate nicer changelogs.

Probably worth having a discussion about at some point. I'll have a quick look around to see how the tools in this space look - I haven't looked into this in a year or so.

Comment (Gudahtt) [reply .circleci/scripts/release-create-gh-release]: After further investigation, it looks like a GitHub App can't interact with the Squash & Merge commit message. There's no way to that commit *after* it was authored but *before* the merge - those happen at the same time right now.

File: .circleci/scripts/release-bump-manifest-version (hunk 7 above)
Comment (whymarrh) [reply .circleci/scripts/release-bump-manifest-version]: Yeah I'm knowingly using a variable that might be unset and setting it to the empty string when that's the case. Since we're explicitly checking the value this isn't an issue.

File: .circleci/scripts/release-bump-manifest-version (hunk 6 above)
Comment (whymarrh) [reply .circleci/scripts/release-bump-manifest-version]: Yeah I believe this is safe since the inline heredoc is buffered but I'll split it into multiple lines to be explicit

File: .circleci/scripts/release-create-gh-release (hunk 8 above)
Comment (whymarrh) [reply .circleci/scripts/release-create-gh-release]: Yes, good point

**Comment by whymarrh** [comment]
Task remaining on this PR:

- [x] Rebase this
- [ ] Add reviewers to the created PR (`@MetaMask/extension-release-team`)
- [ ] Add a commit with the updated changelog headers to the PR
- [ ] Add the changelog contents to body of the GitHub release entry

**Comment by whymarrh** [comment]
> - Add a commit with the updated changelog headers to the PR
> - Add the changelog contents to body of the GitHub release entry

I've added d096ead and 56ddde7 to do exactly those things. I think the code here is ready for review and I'm going to give it a QA pass over on my fork. I'll update here when I've confirmed that this all works as I'd like it to.

File: .circleci/scripts/release-bump-changelog-version [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    exit 1
+fi
+
+version="${CIRCLE_BRANCH/Version-v/}"
+
+if grep --quiet --fixed-strings "$version" CHANGELOG.md
+then
+    printf '%s\n' 'Adding this release to CHANGELOG.md'
+    date_str="$(date '+%a %b %d %Y')"
+    # Replace '## Current Develop Branch\n\n' with "## Current Develop Branch\n\n## ${version} ${date_str}\n\n"
```
Comment (Gudahtt) [inline .circleci/scripts/release-bump-changelog-version]: I'm guessing a command should be here to perform this replacement?

**Review by Gudahtt — CHANGES_REQUESTED** [review]

File: .circleci/scripts/release-bump-changelog-version (hunk 9 above)
Comment (whymarrh) [reply .circleci/scripts/release-bump-changelog-version]: Whoops, yes

Comment (whymarrh) [reply .circleci/scripts/release-bump-changelog-version]: What command would you suggest for this?

Comment (Gudahtt) [reply .circleci/scripts/release-bump-changelog-version]: Maybe something like `sed -i'' '/## Current Develop Branch/c## [replacement title]' CHANGELOG.md `?

Comment (whymarrh) [reply .circleci/scripts/release-bump-changelog-version]: Yeah I have something similar locally but it's not quite correct:

```
sed -i .bak -e "s/## Current Develop Branch/## Current Develop Branch\\n\\n## ${version} ${date_str}/" CHANGELOG.md
```

Comment (whymarrh) [reply .circleci/scripts/release-bump-changelog-version]: I'm exploring solutions using AWK now

Comment (whymarrh) [reply .circleci/scripts/release-bump-changelog-version]: Got a solution with AWK

**Review by Gudahtt — CHANGES_REQUESTED** [review]
I believe `release-bump-changelog-version` needs to be made executable? All the scripts but that one are showing as `+x` for me

**Comment by whymarrh** [comment]
I'm giving this one more QA pass on my fork now, I'll leave a comment here when it's ready for review

**Comment by whymarrh** [comment]
Cool, I'm happy with where this is. I think this is good for another review. As it stands, this successfully creates a GitHub Release with the changelog and assets attached. You can see an example of the process on my fork:

1. The version pull request: https://github.com/whymarrh/metamask-extension/pull/22
2. The commit in particular with the updated manifest and changelog: https://github.com/whymarrh/metamask-extension/pull/22/commits/4933fcdc2812d04e3209ca309619ded580d29356
3. The GitHub Release: [Version 7.0.0](https://github.com/whymarrh/metamask-extension/releases/tag/v7.0.0)

Note that the release contains all the lines in the changelog for the given version so we have the opportunity to add content both _before_ and _after_ the pull request is created.

Note as well that there is a slight but important difference between the name of the branches we currently use and the branches we'll need to use to trigger these scripts: there is a `v` prefix for the version number. Additionally, we should be using [the **Squash and merge** option on the version pull request on GitHub](https://help.github.com/en/articles/about-pull-request-merges#squash-and-merge-your-pull-request-commits) to ensure that the commit message is formatted correctly.

cc @danfinlay @Gudahtt

**Comment by whymarrh** [comment]
I've added fdc2a7e as well to set the reviewer on the PR

**Review by Gudahtt — APPROVED** [review]
LGTM

**Comment by danfinlay** [comment]
Could we add your description of this flow (documentation of how to use these scripts) to our `docs` folder?

I also notice the line item:
- [ ] Set GITHUB_TOKEN and GITHUB_TOKEN_USER environment variables in CircleCI

Should these tokens be the @MetaMaskBot keys? I imagine it would be posting the PRs.

**Comment by whymarrh** [comment]
> I also notice the line item:
>
> - Set GITHUB_TOKEN and GITHUB_TOKEN_USER environment variables in CircleCI
>
> Should these tokens be the @metamaskbot keys? I imagine it would be posting the PRs.

Yes, that'll be set once these are merged in, via the Circle CI config page.

> Could we add your description of this flow (documentation of how to use these scripts) to our docs folder?

Sure can! I'll note here though that these scripts aren't run directly and are only for CI (they've got checks in place to make sure as well) so it'll be more documenting how to trigger these scripts.

**Review by danfinlay — APPROVED** [review]

**Comment by whymarrh** [comment]
I'm going to create a separate PR with some updates to the docs

**Comment by whymarrh** [comment]
> Set `GITHUB_TOKEN` and `GITHUB_TOKEN_USER` environment variables in CircleCI

I've set these on CI:

[](https://circleci.com/gh/MetaMask/metamask-extension/edit#env-vars)
